// yarnlock-dedupe-check/lib/cli.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInputParams = void 0;
const commander_1 = __importDefault(require("commander"));
const pkg = require("../package.json");
const getInputParams = () => {
    const program = new commander_1.default.Command();
    program
        .version(pkg.version)
        .description(pkg.description)
        .option("-i --input <yarn.lock | package-lock.json>", "input yarn.lock file")
        .option("--json <output.json>", "output filename.")
        .option("--html <output.html>", "output html filename.")
        .option("--ignore <regex>", "ignore regex pattern")
        .option("--test <regex>", "test target pattern")
        .option("--warn <regex>", "warning target pattern")
        .parse(process.argv);
    const options = program.opts();
    return {
        inputLockFile: options["input"],
        jsonFileName: options["json"],
        testPattern: options["test"],
        warningPattern: options["warn"],
        ignorePattern: options["ignore"],
        html: options["html"],
    };
};
exports.getInputParams = getInputParams;
