// yarnlock-dedupe-check/lib/cli.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const commander_1 = __importDefault(require("commander"));
const pkg = require("../package.json");
exports.getInputParams = () => {
    commander_1.default
        .version(pkg.version)
        .description(pkg.description)
        .option("-i --input <yarn.lock | package-lock.json>", "input yarn.lock file")
        .option("--json <output.json>", "output filename.")
        .option("--html <output.html>", "output html filename.")
        .option("--ignore <regex>", "ignore regex pattern")
        .option("--test <regex>", "test target pattern")
        .option("--warn <regex>", "warning target pattern")
        .parse(process.argv);
    return {
        inputLockFile: commander_1.default["input"],
        jsonFileName: commander_1.default["json"],
        testPattern: commander_1.default["test"],
        warningPattern: commander_1.default["warn"],
        ignorePattern: commander_1.default["ignore"],
        html: commander_1.default["html"],
    };
};
